### PR TITLE
[ios/tvos] ARM64 simulator JIT Hack and CMAKE_SYSTEM_VARIANT  iOSSimulator

### PR DESCRIPTION
**Pull request by lambdageek**
On iossimulator-arm64 and tvossimulator-arm64, use the `pthread_jit_write_protect_np` function to toggle the W^X bit for the current thread.

Unfortunately, the above function is marked as `__API_UNAVAILABLE` on ios, tvos, watchos in the apple headers.  But in the simulator frameworks it is still present and usable.  So we declare our own prototype of the function and avoid the clang error.

---

This reverts commit 89a816f, but for iOS and tvOS simulators, not for Catalyst.

That commit in turn reverted a9f1207

(ie this current commit is the same as a9f1207 but for ios simulator instead of catalyst)

---

Build system changes:

Set `CMAKE_SYSTEM_VARIANT` to "iOSSimulator" or "tvOSSimulator".

In Mono, set `CMAKE_DARWIN_SIMULATOR`

### CODE REVIEW

**Review by steveisok — APPROVED** [review]

File: src/mono/mono/utils/write-protect.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
+ * \file
+ */
+
+#ifndef __MONO_WRITE_PROTECT_H__
+#define __MONO_WRITE_PROTECT_H__
+
+#include <mono/utils/mono-publib.h>
+
+#if defined (HOST_MACCAT) && defined (__aarch64__)
```
Comment (CoffeeFlux) [inline src/mono/mono/utils/write-protect.h:10]: I thought this was supposed to not only apply to catalyst?

**Review by CoffeeFlux — COMMENTED** [review]
The repeated reversions are confusing: can you just give me a quick description of what configurations require the hack and why?

File: src/mono/mono/utils/write-protect.h (hunk 1 above)
Comment (imhameed) [reply src/mono/mono/utils/write-protect.h:10]: Agreed; wouldn't you want `if defined(HOST_DARWIN_SIMULATOR)`?

Comment (lambdageek) [reply src/mono/mono/utils/write-protect.h:10]: oops. you're right

**Comment by lambdageek** [comment]
> The repeated reversions are confusing: can you just give me a quick description of what configurations require the hack and why?

Updated the commit message and the PR description.  We're only enabling this for iossimulator-arm64 and tvossimulator-arm64. The reasoning being is that those are meant to run just on the local development machine, so there is no risk of running afoul of any App Store rules.

File: src/mono/mono/utils/mono-codeman.c [hunk 2]
```diff
@@ -668,9 +669,11 @@ mono_codeman_enable_write (void)
 		mono_native_tls_set_value (write_level_tls_id, GINT_TO_POINTER (level));
 		pthread_jit_write_protect_np (0);
```
Comment (imhameed) [inline src/mono/mono/utils/mono-codeman.c:670]: This isn't a part of this diff, but: is the Apple-supplied definition of `pthread_jit_write_protect_np` defined with `__attribute__((weak_import))`? If not, do we expect that Mono built on a macOS 11+ amd64 machine will fail during symbol resolution on older OS versions?

Comment (lambdageek) [reply src/mono/mono/utils/mono-codeman.c:670]: hm. good catch.

```
__attribute__((availability(macos,introduced=11.0)))
__attribute__((availability(ios,unavailable))) __attribute__((availability(tvos,unavailable))) __attribute__((availability(watchos,unavailable))) __attribute__((availability(driverkit,unavailable)))
void pthread_jit_write_protect_np(int enabled);
```

```
$ nm foo
0000000100008008 d __dyld_private
0000000100000000 T __mh_execute_header
0000000100003f80 T _main
                 U _pthread_jit_write_protect_np
                 U dyld_stub_binder
```

Doesn't look weak.  So yea we should probably not build this on amd64 at all.

Comment (lambdageek) [reply src/mono/mono/utils/mono-codeman.c:670]: Hmm... actually if I compile my toy with `-mmacosx-version-min=10.5` then  `nm -m` is better:

```
$ nm -m foo
0000000100004110 (__DATA,__bss) non-external _AvailabilityVersionCheck
                 (undefined) external _CFBundleGetVersionNumber (from CoreFoundation)
00000001000040f0 (__DATA,__bss) non-external _CompatibilityDispatchOnceCounter
0000000100004108 (__DATA,__bss) non-external _DispatchOnceCounter
00000001000040f8 (__DATA,__bss) non-external _GlobalMajor
00000001000040fc (__DATA,__bss) non-external _GlobalMinor
0000000100004100 (__DATA,__bss) non-external _GlobalSubminor
00000001000040d0 (__DATA,__common) external _NXArgc
00000001000040d8 (__DATA,__common) external _NXArgv
                 (undefined) external ___assert_rtn (from libSystem)
0000000100003360 (__TEXT,__text) non-external (was a private external) ___clang_at_available_requires_core_foundation_framework
0000000100003370 (__TEXT,__text) non-external (was a private external) ___isOSVersionAtLeast
0000000100003520 (__TEXT,__text) non-external (was a private external) ___isPlatformOrVariantPlatformVersionAtLeast
0000000100003a40 (__TEXT,__text) non-external ___isPlatformOrVariantPlatformVersionAtLeast.cold.1
0000000100003400 (__TEXT,__text) non-external (was a private external) ___isPlatformVersionAtLeast
00000001000040e8 (__DATA,__common) external ___progname
                 (undefined) external ___stack_chk_fail (from libSystem)
                 (undefined) external ___stack_chk_guard (from libSystem)
00000001000032f0 (__TEXT,__text) non-external (was a private external) __dyld_func_lookup
0000000100003700 (__TEXT,__text) non-external __initializeAvailabilityCheck
0000000100000000 (absolute) [referenced dynamically] external __mh_execute_header
00000001000033f0 (__TEXT,__text) non-external _compatibilityInitializeAvailabilityCheck
                 (undefined) weak external _dispatch_once_f (from libSystem)
                 (undefined) external _dlsym (from libSystem)
00000001000040e0 (__DATA,__common) external _environ
                 (undefined) external _exit (from libSystem)
                 (undefined) external _fclose (from libSystem)
                 (undefined) external _fopen (from libSystem)
                 (undefined) external _fread (from libSystem)
                 (undefined) external _free (from libSystem)
                 (undefined) external _fseek (from libSystem)
                 (undefined) external _ftell (from libSystem)
                 (undefined) weak external _gaz (from libgaz)
0000000100003510 (__TEXT,__text) non-external _initializeAvailabilityCheck
0000000100003300 (__TEXT,__text) external _main
                 (undefined) external _malloc (from libSystem)
                 (undefined) weak external _pthread_jit_write_protect_np (from libSystem)
                 (undefined) external _rewind (from libSystem)
                 (undefined) external _sscanf (from libSystem)
00000001000032dc (__TEXT,__text) non-external (was a private external) dyld_stub_binding_helper
00000001000032a0 (__TEXT,__text) external start
```

in particular `(undefined) weak external _pthread_jit_write_protect_np (from libSystem)`

So as long as we set `-mmacosx-version-min=` we're okay

Comment (imhameed) [reply src/mono/mono/utils/mono-codeman.c:670]: Ah, __attribute__((availability)) does this. It's even documented here: https://clang.llvm.org/docs/AttributeReference.html

File: eng/native/build-commons.sh [hunk 3]
```diff
@@ -94,6 +94,10 @@ build_native()
         cmakeArgs="-DCMAKE_SYSTEM_VARIANT=MacCatalyst $cmakeArgs"
     fi
 
+    if [[ "$targetOS" == "iOSSimulator" || "$targetOS" == "tvOSSimulator" ]]; then
+	cmakeArgs="-DCMAKE_SYSTEM_VARIANT=$targetOS $cmakeArgs"
```
Comment (akoeplinger) [inline eng/native/build-commons.sh:98]: nit: spaces for indentation:
```suggestion
        cmakeArgs="-DCMAKE_SYSTEM_VARIANT=$targetOS $cmakeArgs"
```

Comment (akoeplinger) [inline eng/native/build-commons.sh:98]: nit: spaces for indentation:
```suggestion
        cmakeArgs="-DCMAKE_SYSTEM_VARIANT=$targetOS $cmakeArgs"
```

File: src/mono/mono/utils/write-protect.c [hunk 4]
```diff
@@ -0,0 +1,27 @@
+/**
+ * \file
+ */
+
+#include "config.h"
+
+#include "mono-compiler.h"
+#include "write-protect.h"
+
+#if (defined(HOST_IOS) || defined(HOST_TVOS)) && defined (HOST_DARWIN_SIMULATOR) && defined (__aarch64__)
```
Comment (akoeplinger) [inline src/mono/mono/utils/write-protect.c:10]: can we use `HOST_ARM64` instead of the `__aarch64__` compiler macro?

**Review by imhameed — APPROVED** [review]

File: src/mono/mono/utils/mono-codeman.c [hunk 5]
```diff
@@ -683,17 +694,14 @@ mono_codeman_enable_write (void)
 void
 mono_codeman_disable_write (void)
 {
-#ifdef HAVE_PTHREAD_JIT_WRITE_PROTECT_NP
-	if (__builtin_available (macOS 11, *)) {
+#ifdef USE_WRITE_PROTECT
+	if (JIT_WRITE_PROTECT_AVAIL()) {
```
Comment (lambdageek) [inline src/mono/mono/utils/mono-codeman.c:698]:
```suggestion
#ifdef CHECK_WRITE_PROTECT_AVAIL
	if (__builtin_available (macOS 11, *))
#endif
	{
```

**Comment by lambdageek** [comment]
We don't need this anymore.  #53197 added a no-exec mode to the code manager so that cross-compilers don't need to use the pthread write_protect function.  And #51536 will use the "JustInterp" (`--aot=interp` / `MONO_AOT_MODE_INTERP`) to run the apps - which does not require JITable pages for trampolines, etc.

**Pull request closed without merging**